### PR TITLE
[Relay][Frontend][TensorFlow] Support BatchMatMul with input dimensions larger than 3

### PR DESCRIPTION
**Pull request by soiferj**
This change adds support for BatchMatMul with n-dimensional inputs. For example, batch matrix multiplication of (2, 3, 4, 5) x (2, 3, 5, 4).

### CODE REVIEW

**Comment by soiferj** [comment]
@srkreddy1238 would you mind taking a look?

File: python/tvm/relay/frontend/tensorflow.py [hunk 1]
```diff
@@ -448,11 +448,34 @@ def _impl(inputs, attr, params):
...
+        input_y = inputs[1]
+        orig_shape_x = attr['_input_shapes'][input_x]
+        orig_shape_y = attr['_input_shapes'][input_y]
+
+        # reshape n-dimensional batch matmul into 3d
+        if len(orig_shape_x) > 3:
+            outer_dims = [orig_shape_x[i] for i in range(0, len(orig_shape_x) - 2)]
+            num_outer_elts = 1
+            for outer_dim in outer_dims:
+                num_outer_elts *= outer_dim
```
Comment (alexeyr) [inline python/tvm/relay/frontend/tensorflow.py]: Shouldn't this just call `np.prod`?

File: tests/python/frontend/tensorflow/test_forward.py [hunk 2]
```diff
@@ -669,6 +669,10 @@ def test_forward_batch_matmul():
     _test_batch_matmul((3, 5, 4), (3, 4, 5), 'float32', True, True)
     _test_batch_matmul((3, 5, 4), (3, 5, 4), 'int32', True, False)
     _test_batch_matmul((3, 5, 4), (3, 5, 4), 'float32', False, True)
+    _test_batch_matmul((2, 3, 4, 5, 6), (2, 3, 4, 6, 5), 'int32')
+    _test_batch_matmul((2, 3, 4, 5, 6), (2, 3, 4, 6, 5), 'float32', True, True)
+    _test_batch_matmul((2, 3, 4, 5, 6), (2, 3, 4, 5, 6), 'int32', True, False)
+    _test_batch_matmul((2, 3, 4, 5, 6), (2, 3, 4, 5, 6), 'float32', False, True)
```
Comment (alexeyr) [inline tests/python/frontend/tensorflow/test_forward.py]: I'd prefer to have tests with different numbers of outer dimensions.

File: python/tvm/relay/frontend/tensorflow.py (hunk 1 above)
Comment (soiferj) [reply python/tvm/relay/frontend/tensorflow.py]: Updated

File: tests/python/frontend/tensorflow/test_forward.py (hunk 2 above)
Comment (soiferj) [reply tests/python/frontend/tensorflow/test_forward.py]: Updated

**Comment by MarisaKirisame** [comment]
@soiferj how hard is it to make dense/batchmatmul to support more dimension instead?

**Comment by soiferj** [comment]
From looking around the code, it seems like a lot of work. All of computes would have to change, and we would probably want to change all of the schedules as well. We would also have to find and remove all asserts that verify the number of dimensions. Is this a preferred solution? I don't think the reshapes will have much overhead.

**Comment by MarisaKirisame** [comment]
I see. Can you make this function universal then? (move this into relay/op, and allow other ppl to use it as well). I need this function badly.

**Comment by soiferj** [comment]
Sure! I can work on that.

**Comment by soiferj** [comment]
@alexeyr what do you think of the proposed change? Should we make this functionality a little more generic? If so, where do you think it should go?

**Comment by soiferj** [comment]
@marisakirisame would you mind explaining what you want the interface to look like? Because I don’t know how all of the front ends look, I could make a method called ConvertTo3d. Then, converting back to the original dimensions would be handled by the front end (since it’s just one function call).

**Comment by MarisaKirisame** [comment]
@soiferj nevermind, I dont need it anymore, this is good to me.

**Comment by alexeyr** [comment]
@soiferj I'd reply that I can't evaluate at the moment anyway :) But it looks good to me.

**Comment by soiferj** [comment]
@tqchen @tmoreau89 would one of you be able to merge?

**Comment by tqchen** [comment]
Thanks @soiferj !

**Pull request merged**